File: sub_projs/cllm/tools/llm-login/test_offline.py
#!/usr/bin/env python3
"""test_offline.py — llm-login 離線煙霧測試（不連網、不開瀏覽器）。

驗可離線驗的部分：PKCE challenge 正確性、authorize URL 組裝、token store 正規化＋
round-trip、is_expired 邊界、patch cllm config 只動 api_key 不傷其他鍵。真 OAuth 往返
（login/refresh 打真 endpoint）本質要真供應商，屬 WAIT_USER。
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
import tempfile
import urllib.parse

import oauth
import store


def check(cond, msg):
    if not cond:
        raise AssertionError(msg)
    print("  ok  " + msg)


def test_pkce():
    v, c = oauth.gen_pkce()
    want = base64.urlsafe_b64encode(hashlib.sha256(v.encode()).digest()).rstrip(b"=").decode()
    check(c == want, "PKCE challenge = base64url(sha256(verifier)) 無填充")
    check("=" not in c and len(v) >= 43, "verifier 長度≥43、challenge 無 padding")


def test_authorize_url():
    prov = {"authorize_url": "https://p.example/auth", "client_id": "cid",
            "scopes": ["a", "b"], "extra_auth_params": {"prompt": "login"}}
    url = oauth.build_authorize_url(prov, "http://127.0.0.1:8123/callback", "st8", "chal")
    q = dict(urllib.parse.parse_qsl(urllib.parse.urlparse(url).query))
    check(q["code_challenge_method"] == "S256", "帶 S256")
    check(q["scope"] == "a b", "scopes 以空白併接")
    check(q["state"] == "st8" and q["code_challenge"] == "chal", "state/challenge 帶上")
    check(q["prompt"] == "login", "extra_auth_params 併入")


def test_authorize_url_openrouter():
    prov = {"authorize_url": "https://openrouter.ai/auth"}
    url = oauth.build_authorize_url_openrouter(prov, "http://127.0.0.1:8123/callback", "chal")
    q = dict(urllib.parse.parse_qsl(urllib.parse.urlparse(url).query))
    check(q["callback_url"] == "http://127.0.0.1:8123/callback", "OpenRouter 用 callback_url")
    check(q["code_challenge_method"] == "S256" and q["code_challenge"] == "chal", "帶 PKCE")
    check("client_id" not in q and "state" not in q, "OpenRouter 免 client_id／無 state")


def test_token_store_roundtrip():
    with tempfile.TemporaryDirectory() as d:
        os.environ["LLM_CONFIG_DIR"] = d
        rec = store.save_token({"access_token": "AT1", "expires_in": 3600,
                                "refresh_token": "RT1", "token_type": "Bearer"})
        check(rec["refresh_token"] == "RT1" and rec["expires_at"] > 0, "首存帶 refresh＋算 expires_at")
        # 第二次回應沒帶 refresh_token → 沿用舊的
        rec2 = store.save_token({"access_token": "AT2", "expires_in": 3600})
        check(rec2["refresh_token"] == "RT1", "缺 refresh_token 時沿用舊的（不弄丟續期能力）")
        check(store.load_token_silent()["access_token"] == "AT2", "落檔 round-trip")
        mode = oct(os.stat(store.token_path()).st_mode)[-3:]
        check(mode == "600", "token 檔權限 0600")


def test_is_expired():
    check(store.is_expired({"expires_at": 0}) is True, "過去時間＝過期")
    check(store.is_expired({}) is False, "無 expires_at＝不過期型")
    import time as _t
    check(store.is_expired({"expires_at": _t.time() + 3600}) is False, "一小時後＝未過期")


def test_patch_config_preserves_keys():
    with tempfile.TemporaryDirectory() as d:
        os.environ["LLM_CONFIG_DIR"] = d
        cfg_path = store.cllm_config_path()
        os.makedirs(d, exist_ok=True)
        with open(cfg_path, "w", encoding="utf-8") as f:
            json.dump({"endpoint": "http://x/v1/chat/completions", "model": "m",
                       "timeout_ms": 120000}, f)
        store.patch_cllm("BEARER123")  # 不帶 endpoint/model
        got = json.load(open(cfg_path, encoding="utf-8"))
        check(got["api_key"] == "BEARER123", "api_key 寫入")
        check(got["endpoint"] == "http://x/v1/chat/completions" and got["model"] == "m"
              and got["timeout_ms"] == 120000, "沒帶就不動 endpoint/model、其他鍵原樣保留")
        # preset 帶 endpoint/model 時一併設
        store.patch_cllm("BEARER456", "https://y/v1/chat/completions", "gemini-2.5-flash")
        got = json.load(open(cfg_path, encoding="utf-8"))
        check(got["endpoint"] == "https://y/v1/chat/completions"
              and got["model"] == "gemini-2.5-flash" and got["timeout_ms"] == 120000,
              "帶了就一步設好 endpoint/model，timeout_ms 仍保留")


def test_presets_load():
    here = os.path.dirname(os.path.abspath(__file__))
    pdir = os.path.join(here, "providers")
    names = sorted(f for f in os.listdir(pdir) if f.endswith(".json"))
    check(len(names) >= 4, "至少四個 preset：%s" % names)
    for n in names:
        os.environ["LLM_OAUTH_PROVIDER"] = os.path.join(pdir, n)
        prov = store.load_provider()  # JSON 合法＋必填欄位齊（openrouter 免 client_id）
        check(bool(prov.get("cllm_endpoint")), "%s 帶 cllm_endpoint" % n)
    del os.environ["LLM_OAUTH_PROVIDER"]


def main():
    for t in (test_pkce, test_authorize_url, test_authorize_url_openrouter,
              test_token_store_roundtrip, test_is_expired,
              test_patch_config_preserves_keys, test_presets_load):
        print(t.__name__)
        t()
    print("\n全部離線煙霧測試通過。")


if __name__ == "__main__":
    main()

File: sub_projs/cllm/tools/llm-login/test_offline.cpp
// test_offline.cpp — llm-login 離線煙霧測試（不連網、不開瀏覽器、不起 server）。
//
// 測確定性的部分：crypto（SHA-256 已知答案、base64url、PKCE 自洽）、authorize URL 組裝、
// config patch 保留其餘鍵、token 記錄 expires_at 計算。跑：build/tools/llm-login-test-offline。

#include <cstdio>
#include <string>

#include "crypto.hpp"
#include "oauth.hpp"
#include "store.hpp"

static int g_fail = 0;
static void check(const char *name, bool ok) {
  std::printf("  %s %s\n", ok ? " ok " : "FAIL", name);
  if (!ok) ++g_fail;
}

static std::string hex(const std::vector<uint8_t> &b) {
  static const char *h = "0123456789abcdef";
  std::string s;
  for (uint8_t c : b) {
    s.push_back(h[c >> 4]);
    s.push_back(h[c & 15]);
  }
  return s;
}

static void test_sha256() {
  std::printf("• sha256\n");
  // NIST 已知答案
  check("sha256(\"abc\")",
        hex(login::crypto::sha256("abc")) ==
            "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad");
  check("sha256(\"\")",
        hex(login::crypto::sha256("")) ==
            "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855");
}

static void test_base64url() {
  std::printf("• base64url\n");
  auto b = [](std::string s) { return std::vector<uint8_t>(s.begin(), s.end()); };
  check("\"Man\"→TWFu", login::crypto::base64url_nopad(b("Man")) == "TWFu");
  check("\"Ma\"→TWE（無 padding）", login::crypto::base64url_nopad(b("Ma")) == "TWE");
  check("\"M\"→TQ（無 padding）", login::crypto::base64url_nopad(b("M")) == "TQ");
}

static void test_pkce() {
  std::printf("• pkce\n");
  auto [verifier, challenge] = login::crypto::gen_pkce();
  check("verifier 43 字元（base64url of 32B）", verifier.size() == 43);
  check("challenge==base64url(sha256(verifier))",
        challenge == login::crypto::base64url_nopad(login::crypto::sha256(verifier)));
  check("challenge 43 字元", challenge.size() == 43);
  auto [v2, c2] = login::crypto::gen_pkce();
  check("兩次不同（隨機）", verifier != v2);
}

static void test_authorize_url() {
  std::printf("• authorize_url\n");
  login::oauth::J prov;
  glz::read_json(prov, R"({"authorize_url":"https://auth.example/authorize",
    "client_id":"cid123","scope":"read write","token_url":"https://auth.example/token"})");
  std::string url = login::oauth::authorize_url(prov, "http://127.0.0.1:8123/callback", "st8",
                                                "chal");
  auto has = [&](const char *s) { return url.find(s) != std::string::npos; };
  check("response_type=code", has("response_type=code"));
  check("client_id 帶入", has("client_id=cid123"));
  check("redirect_uri URL 編碼", has("redirect_uri=http%3A%2F%2F127.0.0.1%3A8123%2Fcallback"));
  check("scope 編碼（空白→%20）", has("scope=read%20write"));
  check("state 帶入", has("state=st8"));
  check("S256", has("code_challenge_method=S256"));

  login::oauth::Redirect rd = login::oauth::redirect_of(prov);
  check("redirect 預設 127.0.0.1:8123/callback",
        rd.host == "127.0.0.1" && rd.port == 8123 && rd.path == "/callback");
}

static void test_config_patch() {
  std::printf("• config patch\n");
  std::string tmp = "/tmp/llm_login_test_cfg.json";
  login::store::write_file(tmp, R"({"api_key":"old","timeout_ms":5000})", false);
  login::store::patch_cllm(tmp, "sk-new", "http://x/y", "m1");
  login::store::J back = login::store::read_json(tmp);
  auto s = [&](const char *k) {
    return back.contains(k) && back[k].is_string() ? back[k].get_string() : std::string();
  };
  check("api_key 被換", s("api_key") == "sk-new");
  check("endpoint 寫入", s("endpoint") == "http://x/y");
  check("model 寫入", s("model") == "m1");
  check("其餘鍵原樣保留（timeout_ms）",
        back.contains("timeout_ms") && back["timeout_ms"].is_number() &&
            back["timeout_ms"].get_number() == 5000);
  std::remove(tmp.c_str());
}

static void test_token_record() {
  std::printf("• token record\n");
  login::store::J raw;
  glz::read_json(raw, R"({"access_token":"a1","refresh_token":"r1","expires_in":3600})");
  login::store::J rec = login::store::make_token_record(raw, 1000000);
  check("expires_at = now+expires_in",
        rec.contains("expires_at") && rec["expires_at"].get_number() == 1000000 + 3600);
  check("未過期（now）", !login::store::is_expired(rec, 1000000));
  check("已過期（now 超過）", login::store::is_expired(rec, 1000000 + 4000));
  login::store::J norefresh;
  glz::read_json(norefresh, R"({"access_token":"k1"})");
  login::store::J rec2 = login::store::make_token_record(norefresh, 1000000);
  check("無 expires_in → 不過期型永不過期", !login::store::is_expired(rec2, 9999999999));
}

int main() {
  test_sha256();
  test_base64url();
  test_pkce();
  test_authorize_url();
  test_config_patch();
  test_token_record();
  std::printf("\n%s\n", g_fail == 0 ? "全綠 ✅" : "有紅 ❌");
  return g_fail == 0 ? 0 : 1;
}
